In [7]:
import numpy as np
import scanpy as sc

# 1. Single-Cell RNA-seq Dataset Exploration

In [8]:
adata = sc.read(
    filename="s4d8_quality_control.h5ad",
    backup_url="https://figshare.com/ndownloader/files/40014331",
)



In [9]:
# Print summary of the dataset
adata

AnnData object with n_obs × n_vars = 14814 × 20171
    obs: 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_20_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'outlier', 'mt_outlier', 'scDblFinder_score', 'scDblFinder_class'
    var: 'gene_ids', 'feature_types', 'genome', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells'
    layers: 'counts', 'soupX_counts'

In [10]:
# Inspect the Dataset Shape
adata.shape

(14814, 20171)

- Shape: `(14814, 20171)`
  - 14814 cells (rows in `adata.obs`)
  - 20171 genes (columns in `adata.var`)

This tells that there are roughly **14814** single cells and **20171** measured features (genes) in the dataset.

## Observations (Cells)

Each row in `adata.obs` corresponds to a single cell.  
We'll inspect the first few rows to see what metadata fields are present.

In [11]:
# List all columns (metadata fields) in adata.obs
print("Cell metadata fields:", adata.obs_keys())

# Show the first five rows of the cell metadata
adata.obs.head()

Cell metadata fields: ['n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_20_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'outlier', 'mt_outlier', 'scDblFinder_score', 'scDblFinder_class']


n_genes_by_counts  log1p_n_genes_by_counts  total_counts  \
AAACAGCCAAGCTTAT-1               1362                 7.217443        2243.0   
AAACAGCCATAGCTTG-1               2126                 7.662468        4085.0   
AAACAGCCATGTTTGG-1               1340                 7.201171        2227.0   
AAACATGCAACGTGCT-1                828                 6.720220        1211.0   
AAACATGCAATATAGG-1                551                 6.313548         719.0   

                    log1p_total_counts  pct_counts_in_top_20_genes  \
AAACAGCCAAGCTTAT-1            7.716015                    9.719126   
AAACAGCCATAGCTTG-1            8.315322                   10.820073   
AAACAGCCATGTTTGG-1            7.708859                   12.977099   
AAACATGCAACGTGCT-1            7.100027                   14.037985   
AAACATGCAATATAGG-1            6.579251                   15.020862   

                    total_counts_mt  log1p_total_counts_mt  pct_counts_mt  \
AAACAGCCAAGCTTAT-1             18.0               2.944439       0.802497   
AAACAGCCATAGCTTG-1             63.0               4.158883       1.542228   
AAACAGCCATGTTTGG-1             30.0               3.433987       1.347104   
AAACATGCAACGTGCT-1             20.0               3.044523       1.651528   
AAACATGCAATATAGG-1              8.0               2.197225       1.112656   

                    total_counts_ribo  log1p_total_counts_ribo  \
AAACAGCCAAGCTTAT-1              125.0                 4.836282   
AAACAGCCATAGCTTG-1               69.0                 4.248495   
AAACAGCCATGTTTGG-1               42.0                 3.761200   
AAACATGCAACGTGCT-1               44.0                 3.806663   
AAACATGCAATATAGG-1               22.0                 3.135494   

                    pct_counts_ribo  total_counts_hb  log1p_total_counts_hb  \
AAACAGCCAAGCTTAT-1         5.572894              3.0               1.386294   
AAACAGCCATAGCTTG-1         1.689106             10.0               2.397895   
AAACAGCCATGTTTGG-1         1.885945              1.0               0.693147   
AAACATGCAACGTGCT-1         3.633361              6.0               1.945910   
AAACATGCAATATAGG-1         3.059805              1.0               0.693147   

                    pct_counts_hb  outlier  mt_outlier  scDblFinder_score  \
AAACAGCCAAGCTTAT-1       0.133749    False       False           0.036574   
AAACAGCCATAGCTTG-1       0.244798    False       False           0.718944   
AAACAGCCATGTTTGG-1       0.044903    False       False           0.214918   
AAACATGCAACGTGCT-1       0.495458    False       False           0.201699   
AAACATGCAATATAGG-1       0.139082    False       False           0.034913   

                   scDblFinder_class  
AAACAGCCAAGCTTAT-1           singlet  
AAACAGCCATAGCTTG-1           doublet  
AAACAGCCATGTTTGG-1           singlet  
AAACATGCAACGTGCT-1           singlet  
AAACATGCAATATAGG-1           singlet

## Variables (Genes)

Each row in `adata.var` corresponds to a gene (or feature).  
We'll inspect the first few rows to see what annotations (like gene IDs, chromosome info, etc.) are available.

In [12]:
# List all columns (metadata fields) in adata.var
print("Gene metadata fields:", adata.var_keys())

# Show the first five rows of the gene metadata
adata.var.head()

Gene metadata fields: ['gene_ids', 'feature_types', 'genome', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells']


gene_ids    feature_types  genome     mt   ribo     hb  \
AL627309.1  ENSG00000238009  Gene Expression  GRCh38  False  False  False   
AL627309.5  ENSG00000241860  Gene Expression  GRCh38  False  False  False   
LINC01409   ENSG00000237491  Gene Expression  GRCh38  False  False  False   
LINC01128   ENSG00000228794  Gene Expression  GRCh38  False  False  False   
LINC00115   ENSG00000225880  Gene Expression  GRCh38  False  False  False   

            n_cells_by_counts  mean_counts  log1p_mean_counts  \
AL627309.1                 30     0.001831           0.001829   
AL627309.5                146     0.008976           0.008936   
LINC01409                 883     0.063009           0.061104   
LINC01128                 533     0.036140           0.035503   
LINC00115                  72     0.004370           0.004360   

            pct_dropout_by_counts  total_counts  log1p_total_counts  n_cells  
AL627309.1              99.822842          31.0            3.465736       27  
AL627309.5              99.137829         152.0            5.030438      126  
LINC01409               94.785638        1067.0            6.973543      758  
LINC01128               96.852486         612.0            6.418365      437  
LINC00115               99.574820          74.0            4.317488       64

## Layers

Sometimes, an `AnnData` object contains multiple expression layers.  
We can take a look of these layer.

In [13]:
print("Available layers:", list(adata.layers.keys()))

Available layers: ['counts', 'soupX_counts']


In [14]:
# Inspect a small slice of 'counts' layer 
print("First 5 cells x 5 genes in 'counts' layer:\n", adata.layers["counts"][:5, :5].toarray())

First 5 cells x 5 genes in 'counts' layer:
 [[0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]]


In [15]:
# Inspect a small slice of 'soupX_counts' layer 
print("First 5 cells x 5 genes in 'soupX_counts' layer:\n", adata.layers["soupX_counts"][:5, :5].toarray())

First 5 cells x 5 genes in 'soupX_counts' layer:
 [[0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]]


# 2. Data Preprocessing

## Quality Control

In this section, we will:

1. **Calculate QC metrics**: Total counts, number of detected genes, mitochondrial/ribosomal content, etc.
2. **Filter cells**: Remove cells with too few genes or too many genes (potential doublets), and cells with high mitochondrial gene percentages.
3. **Visualize** the effect of filtering.

### 1. Calculate QC Metrics

We'll use `sc.pp.calculate_qc_metrics`, which can automatically compute:
- `n_genes_by_counts` (number of genes per cell)
- `total_counts` (sum of UMIs per cell)
- `adata.var` as `mt`(Mitochondrial percentages) 

In [16]:
sc.pp.calculate_qc_metrics(
    adata,
    qc_vars=['mt', 'ribo', 'hb'],  # adjust to what you have in your var
    inplace=True
)

print("QC metrics added to adata.obs:")
print(adata.obs_keys())
adata.obs.head()


QC metrics added to adata.obs:
['n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_20_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'outlier', 'mt_outlier', 'scDblFinder_score', 'scDblFinder_class', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes']


n_genes_by_counts  log1p_n_genes_by_counts  total_counts  \
AAACAGCCAAGCTTAT-1               1272                 7.149132        2091.0   
AAACAGCCATAGCTTG-1               1783                 7.486613        3494.0   
AAACAGCCATGTTTGG-1               1217                 7.104965        2016.0   
AAACATGCAACGTGCT-1                778                 6.658011        1141.0   
AAACATGCAATATAGG-1                528                 6.270988         689.0   

                    log1p_total_counts  pct_counts_in_top_20_genes  \
AAACAGCCAAGCTTAT-1            7.645876                    9.719126   
AAACAGCCATAGCTTG-1            8.159089                   10.820073   
AAACAGCCATGTTTGG-1            7.609367                   12.977099   
AAACATGCAACGTGCT-1            7.040536                   14.037985   
AAACATGCAATATAGG-1            6.536692                   15.020862   

                    total_counts_mt  log1p_total_counts_mt  pct_counts_mt  \
AAACAGCCAAGCTTAT-1             17.0               2.890372       0.813008   
AAACAGCCATAGCTTG-1             56.0               4.043051       1.602748   
AAACAGCCATGTTTGG-1             28.0               3.367296       1.388889   
AAACATGCAACGTGCT-1             19.0               2.995732       1.665206   
AAACATGCAATATAGG-1              8.0               2.197225       1.161103   

                    total_counts_ribo  log1p_total_counts_ribo  ...  \
AAACAGCCAAGCTTAT-1              123.0                 4.820282  ...   
AAACAGCCATAGCTTG-1               61.0                 4.127134  ...   
AAACAGCCATGTTTGG-1               35.0                 3.583519  ...   
AAACATGCAACGTGCT-1               42.0                 3.761200  ...   
AAACATGCAATATAGG-1               19.0                 2.995732  ...   

                    log1p_total_counts_hb  pct_counts_hb  outlier  mt_outlier  \
AAACAGCCAAGCTTAT-1               1.386294       0.143472    False       False   
AAACAGCCATAGCTTG-1               2.302585       0.257584    False       False   
AAACAGCCATGTTTGG-1               0.693147       0.049603    False       False   
AAACATGCAACGTGCT-1               1.945910       0.525855    False       False   
AAACATGCAATATAGG-1               0.000000       0.000000    False       False   

                    scDblFinder_score  scDblFinder_class  \
AAACAGCCAAGCTTAT-1           0.036574            singlet   
AAACAGCCATAGCTTG-1           0.718944            doublet   
AAACAGCCATGTTTGG-1           0.214918            singlet   
AAACATGCAACGTGCT-1           0.201699            singlet   
AAACATGCAATATAGG-1           0.034913            singlet   

                    pct_counts_in_top_50_genes pct_counts_in_top_100_genes  \
AAACAGCCAAGCTTAT-1                   16.834051                   25.298900   
AAACAGCCATAGCTTG-1                   18.717802                   26.617058   
AAACAGCCATGTTTGG-1                   20.882937                   29.464286   
AAACATGCAACGTGCT-1                   22.874671                   32.602980   
AAACATGCAATATAGG-1                   24.818578                   37.880987   

                    pct_counts_in_top_200_genes  pct_counts_in_top_500_genes  
AAACAGCCAAGCTTAT-1                    37.589670                    63.079866  
AAACAGCCATAGCTTG-1                    37.893532                    58.643389  
AAACAGCCATGTTTGG-1                    41.468254                    64.434524  
AAACATGCAACGTGCT-1                    49.342682                    75.635408  
AAACATGCAATATAGG-1                    52.394775                    95.936139  

[5 rows x 22 columns]